src/bls12_381/ec.rs

```diff
@@ -194,6 +245,26 @@ macro_rules! curve_impl {
                 (*self).into()
             }
 
+            fn hash(seed: &[u8], _nonce: &[u8]) -> Self {
+                // The construction of Foque et al. requires us to construct two
+                // "random oracles" in the field, encode their image with `sw_encode`,
+                // and finally add them.
+                // This is obtained simply by prepending to the nonce ofFq::hash 0x00 and 0xff respectively.
+                // Note: the nonce is assumed to be at most 32 bytes, any extra data will be discarded.
+                let len = cmp::min(32, _nonce.len());
+                let nonce = &mut [0u8; 33];
+                nonce[.. len].copy_from_slice(&_nonce[.. len]);
+
+                // nonce[32] = 0x00;
+                let t1 = Self::Base::hash(seed, nonce);
+                let t1 = Self::sw_encode(t1);
+
+                nonce[32] = 0xff;
+                let t2 = Self::Base::hash(seed, nonce);
+                let mut t2 = Self::sw_encode(t2);
+                t2.add_assign(&t1);
```

Comment:
This seems to generate points in the wrong group. Add a loop into `test_hash`. 

https://github.com/herumi/mcl/blob/9fbc54305d01b984e39d83e96bfa94bb17648a86/include/mcl/bn.hpp#L76 uses sw_encode result directly, no adding.